### PR TITLE
Tidy up PlainListenableActionFuture

### PR DESCRIPTION
**Pull request by DaveCTurner**
As of #53261 today `PlainListenableActionFuture` is the sole
implementation of `ListenableActionFuture`, the reason for the
`Plain...` prefix is now lost. It also heavily uses variable shadowing
in its implementation which makes it quite hard to read, uses a mix of
`volatile` fields and mutexes, and keeps hold of the deferred listeners
even after completion. Finally it does not have many tests.

This commit drops the unnecessary interface and renames the class to
simply `ListenableActionFuture`. It reworks the implementation to avoid
shadowing variables, drops the deferred listeners on completion, and
moves to an entirely mutex-based implementation. Finally, it adds
another test that it works as advertised.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/support/PlainListenableActionFuture.java [hunk 1]
```diff
@@ -1,99 +0,0 @@
...
-        internalAddListener(listener);
-    }
-
-    @Override
-    protected void done(boolean success) {
-        super.done(success);
-        synchronized (this) {
-            executedListeners = true;
-        }
-        Object listeners = this.listeners;
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/support/PlainListenableActionFuture.java:46]: This is the only volatile read outside of the mutex, but we just had the mutex anyway, may as well move this up a line and drop the `volatile`.

File: server/src/main/java/org/elasticsearch/action/support/PlainListenableActionFuture.java [hunk 2]
```diff
@@ -1,99 +0,0 @@
...
-        return listenerResponse;
-    }
-
-    private void internalAddListener(ActionListener<T> listener) {
-        boolean executeImmediate = false;
-        synchronized (this) {
-            if (executedListeners) {
-                executeImmediate = true;
-            } else {
-                Object listeners = this.listeners;
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/support/PlainListenableActionFuture.java:70]: Reworked this to have better names and to use `final` variables throughout, which introduces one extra local variable but makes it all a lot clearer.

File: server/src/main/java/org/elasticsearch/action/support/PlainListenableActionFuture.java [hunk 3]
```diff
@@ -1,99 +0,0 @@
...
-     * @param <T> the result of the future
-     * @return a listenable future
-     */
-    public static <T> PlainListenableActionFuture<T> newListenableFuture() {
-        return new PlainListenableActionFuture<>();
-    }
-
-    @Override
-    public void addListener(final ActionListener<T> listener) {
-        internalAddListener(listener);
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/support/PlainListenableActionFuture.java:37]: The only caller of `internalAddListener`, inlined it.

**Review by henningandersen — APPROVED** [review]
Nice cleanup, LGTM.

**Pull request merged**